File: etrike_protocol/codecs/typescript/ses.ts
import { compute, verify } from "../../profiles/xor8_ff_v1";
import { type CanFrame, type CodecStatus, type DecodeResult, type EncodeResult, frame, validateFrame } from "./types";

export const SES_BUS = "low";
export const SES_DLC = 8;
export const SES_COMMAND_ID = 0x169;
export const SES_STATUS_ID = 0x201;
export const SES_ERROR_INFO_ID = 0x202;
export const SES_VERSION_ID = 0x203;
export const SES_TEST_ID = 0x6fa;

function validate(input: CanFrame, id: number, checksum = false): CodecStatus {
  const status = validateFrame(input, { bus: SES_BUS, id, frameFormat: "standard", dlc: SES_DLC });
  if (status !== "ok") return status;
  if (checksum && !verify(input.data.subarray(0, 7), input.data[7])) return "checksum_mismatch";
  return "ok";
}

function integer(value: unknown, fallback: number): number | undefined {
  const selected = value ?? fallback;
  return typeof selected === "number" && Number.isInteger(selected) ? selected : undefined;
}

export function encodeSesCommand(values: Readonly<Record<string, unknown>>): EncodeResult {
  if (
    (values.alignment_enable !== undefined && typeof values.alignment_enable !== "boolean") ||
    (values.control_enable !== undefined && typeof values.control_enable !== "boolean")
  ) return ["value_out_of_range"];
  const targetAngle = integer(values.target_angle_raw, 0);
  const targetSpeed = integer(values.target_speed_raw, 328);
  const counter = integer(values.rolling_counter, 0);
  const vehicleSpeed = integer(values.vehicle_speed_raw, 0);
  if (targetAngle === undefined || targetSpeed === undefined || counter === undefined || vehicleSpeed === undefined) {
    return ["value_out_of_range"];
  }
  if (targetAngle < -0x8000 || targetAngle > 0x7fff || targetSpeed < 125 || targetSpeed > 525) {
    return ["value_out_of_range"];
  }
  if (counter < 0 || counter > 15 || vehicleSpeed < 0 || vehicleSpeed > 0xff) return ["value_out_of_range"];
  const payload = new Uint8Array(SES_DLC);
  payload[0] = (values.alignment_enable === true ? 1 : 0) | (values.control_enable === true ? 2 : 0);
  new DataView(payload.buffer).setInt16(2, targetAngle, true);
  payload[4] = targetSpeed & 0xff;
  payload[5] = 0x03 | ((targetSpeed >> 6) & 0x0c) | (counter << 4);
  payload[6] = vehicleSpeed;
  payload[7] = compute(payload.subarray(0, 7));
  return ["ok", frame(SES_BUS, SES_COMMAND_ID, "standard", payload)];
}

export function decodeSesCommand(input: CanFrame): DecodeResult<Record<string, unknown>> {
  const status = validate(input, SES_COMMAND_ID, true);
  if (status !== "ok") return [status];
  if ((input.data[5] & 0x03) !== 0x03) return ["constant_mismatch"];
  const targetSpeed = input.data[4] | ((input.data[5] & 0x0c) << 6);
  if (targetSpeed < 125 || targetSpeed > 525) return ["value_out_of_range"];
  return ["ok", {
    alignment_enable: (input.data[0] & 1) !== 0,
    control_enable: (input.data[0] & 2) !== 0,
    target_angle_raw: new DataView(input.data.buffer, input.data.byteOffset, input.data.byteLength).getInt16(2, true),
    target_speed_raw: targetSpeed,
    rolling_counter: input.data[5] >> 4,
    vehicle_speed_raw: input.data[6],
  }];
}

export function decodeSesStatus(input: CanFrame): DecodeResult<Record<string, unknown>> {
  const status = validate(input, SES_STATUS_ID, true);
  if (status !== "ok") return [status];
  const view = new DataView(input.data.buffer, input.data.byteOffset, input.data.byteLength);
  return ["ok", {
    angle_aligned: (input.data[0] & 1) !== 0,
    control_mode: (input.data[0] >> 1) & 3,
    error_status: (input.data[0] >> 6) & 3,
    steering_angle_raw: view.getUint16(2, true),
    target_angle_speed_raw: view.getInt16(4, true),
    steering_torque_raw: input.data[5],
    rolling_counter_enabled: (input.data[6] & 1) !== 0,
    checksum_enabled: (input.data[6] & 2) !== 0,
    rolling_counter: input.data[6] >> 4,
  }];
}

export function decodeSesErrorInfo(input: CanFrame): DecodeResult<{ raw: Uint8Array }> {
  const status = validate(input, SES_ERROR_INFO_ID);
  return status === "ok" ? ["ok", { raw: input.data.slice() }] : [status];
}

export function decodeSesVersion(input: CanFrame): DecodeResult<{ raw: Uint8Array }> {
  const status = validate(input, SES_VERSION_ID);
  return status === "ok" ? ["unsupported_semantics", { raw: input.data.slice() }] : [status];
}

export const decodeSesVersionRaw = decodeSesVersion;

export function decodeSesTest(input: CanFrame): DecodeResult<Record<string, number>> {
  const status = validate(input, SES_TEST_ID);
  if (status !== "ok") return [status];
  const view = new DataView(input.data.buffer, input.data.byteOffset, input.data.byteLength);
  return ["ok", {
    motor_current_raw: view.getInt16(1, true),
    ecu_temperature_raw: view.getUint16(3, true),
    supply_voltage_raw: view.getUint16(5, true),
  }];
}

File: etrike_protocol/codecs/python/ses.py
from __future__ import annotations

from collections.abc import Mapping

from protocol.profiles.xor8_ff_v1 import compute, verify

from .types import CodecStatus, Frame, validate_frame

BUS = "low"
DLC = 8
COMMAND_ID = 0x169
STATUS_ID = 0x201
ERROR_INFO_ID = 0x202
VERSION_ID = 0x203
TEST_ID = 0x6FA


def _validate(frame: Frame, can_id: int, *, checksum: bool = False) -> CodecStatus:
    status = validate_frame(frame, bus=BUS, can_id=can_id, frame_format="standard", dlc=DLC)
    if status != "ok":
        return status
    if checksum and not verify(frame.data[:7], frame.data[7]):
        return "checksum_mismatch"
    return "ok"


def _integer(value: object, default: int) -> int | None:
    selected = default if value is None else value
    return selected if isinstance(selected, int) and not isinstance(selected, bool) else None


def encode_command(values: Mapping[str, object]) -> tuple[CodecStatus, Frame | None]:
    alignment_enable = values.get("alignment_enable", False)
    control_enable = values.get("control_enable", False)
    target_angle = _integer(values.get("target_angle_raw"), 0)
    target_speed = _integer(values.get("target_speed_raw"), 328)
    counter = _integer(values.get("rolling_counter"), 0)
    vehicle_speed = _integer(values.get("vehicle_speed_raw"), 0)
    if not isinstance(alignment_enable, bool) or not isinstance(control_enable, bool):
        return "value_out_of_range", None
    if target_angle is None or target_speed is None or counter is None or vehicle_speed is None:
        return "value_out_of_range", None
    if not -0x8000 <= target_angle <= 0x7FFF or not 125 <= target_speed <= 525:
        return "value_out_of_range", None
    if not 0 <= counter <= 15 or not 0 <= vehicle_speed <= 0xFF:
        return "value_out_of_range", None
    payload = bytearray(DLC)
    payload[0] = int(alignment_enable) | (int(control_enable) << 1)
    payload[2:4] = target_angle.to_bytes(2, "little", signed=True)
    payload[4] = target_speed & 0xFF
    payload[5] = 0x03 | ((target_speed >> 6) & 0x0C) | (counter << 4)
    payload[6] = vehicle_speed
    payload[7] = compute(payload[:7])
    return "ok", Frame(BUS, COMMAND_ID, "standard", payload)


def decode_command(frame: Frame) -> tuple[CodecStatus, dict[str, object] | None]:
    status = _validate(frame, COMMAND_ID, checksum=True)
    if status != "ok":
        return status, None
    if frame.data[5] & 0x03 != 0x03:
        return "constant_mismatch", None
    target_speed = frame.data[4] | ((frame.data[5] & 0x0C) << 6)
    if not 125 <= target_speed <= 525:
        return "value_out_of_range", None
    return "ok", {
        "alignment_enable": bool(frame.data[0] & 0x01),
        "control_enable": bool(frame.data[0] & 0x02),
        "target_angle_raw": int.from_bytes(frame.data[2:4], "little", signed=True),
        "target_speed_raw": target_speed,
        "rolling_counter": frame.data[5] >> 4,
        "vehicle_speed_raw": frame.data[6],
    }


def decode_status(frame: Frame) -> tuple[CodecStatus, dict[str, object] | None]:
    status = _validate(frame, STATUS_ID, checksum=True)
    if status != "ok":
        return status, None
    return "ok", {
        "angle_aligned": bool(frame.data[0] & 0x01),
        "control_mode": (frame.data[0] >> 1) & 0x03,
        "error_status": (frame.data[0] >> 6) & 0x03,
        "steering_angle_raw": int.from_bytes(frame.data[2:4], "little"),
        "target_angle_speed_raw": int.from_bytes(frame.data[4:6], "little", signed=True),
        "steering_torque_raw": frame.data[5],
        "rolling_counter_enabled": bool(frame.data[6] & 0x01),
        "checksum_enabled": bool(frame.data[6] & 0x02),
        "rolling_counter": frame.data[6] >> 4,
    }


def decode_error_info(frame: Frame) -> tuple[CodecStatus, dict[str, bytes] | None]:
    status = _validate(frame, ERROR_INFO_ID)
    return (status, {"raw": frame.data} if status == "ok" else None)


def decode_version(frame: Frame) -> tuple[CodecStatus, dict[str, bytes] | None]:
    status = _validate(frame, VERSION_ID)
    if status != "ok":
        return status, None
    return "unsupported_semantics", {"raw": frame.data}


decode_version_raw = decode_version


def decode_test(frame: Frame) -> tuple[CodecStatus, dict[str, int] | None]:
    status = _validate(frame, TEST_ID)
    if status != "ok":
        return status, None
    return "ok", {
        "motor_current_raw": int.from_bytes(frame.data[1:3], "little", signed=True),
        "ecu_temperature_raw": int.from_bytes(frame.data[3:5], "little"),
        "supply_voltage_raw": int.from_bytes(frame.data[5:7], "little"),
    }
